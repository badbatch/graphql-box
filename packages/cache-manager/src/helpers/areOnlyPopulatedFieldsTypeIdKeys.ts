import type { PlainData } from '@graphql-box/core';
import { isArray, isPlainObject } from '@graphql-box/helpers';
import { isNumber, isString } from 'lodash-es';

const checkValue = (value: unknown, typeIDKey: string): boolean => {
  if (isArray(value)) {
    return value.reduce<boolean>((acc, entry) => {
      if (!acc) {
        return false;
      }

      return checkValue(entry, typeIDKey);
    }, true);
  }

  if (isPlainObject(value)) {
    return recursivelyCheckProps(value, typeIDKey);
  }

  return false;
};

const recursivelyCheckProps = (data: PlainData, typeIDKey: string): boolean => {
  const keys = isPlainObject(data) ? Object.keys(data) : [...data.keys()];

  if (keys.length === 1 && isPlainObject(data) && !!data[typeIDKey]) {
    return true;
  }

  return keys.reduce<boolean>((acc, key) => {
    if (!acc) {
      return false;
    }

    if (isNumber(key) && isArray(data)) {
      return checkValue(data[key], typeIDKey);
    } else if (isString(key) && isPlainObject(data)) {
      return checkValue(data[key], typeIDKey);
    }

    return acc;
  }, true);
};

export const areOnlyPopulatedFieldsTypeIdKeys = (data: PlainData, typeIDKey: string) => {
  return recursivelyCheckProps(data, typeIDKey);
};
