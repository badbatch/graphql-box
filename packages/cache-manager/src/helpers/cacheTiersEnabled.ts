import { type CacheTiersEnabled } from '../types.ts';

export const allCacheTiersDisabled = ({ entity, queryResponse, requestPath }: CacheTiersEnabled) =>
  !entity && !queryResponse && !requestPath;

export const entityCacheTierEnabled = (tiersEnabled: CacheTiersEnabled) => !!tiersEnabled.entity;

export const entityOrRequestPathCacheTiersEnabled = ({ entity, requestPath }: CacheTiersEnabled) =>
  !!(entity ?? requestPath);

export const queryResponseCacheTierEnabled = (tiersEnabled: CacheTiersEnabled) => !!tiersEnabled.queryResponse;

export const requestPathCacheTierEnabled = (tiersEnabled: CacheTiersEnabled) => !!tiersEnabled.requestPath;

export const someCacheTiersEnabled = (tiersEnabled: CacheTiersEnabled) => !allCacheTiersDisabled(tiersEnabled);
