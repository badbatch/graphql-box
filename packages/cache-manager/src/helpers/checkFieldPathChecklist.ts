import { type CheckFieldPathChecklistResult, type FieldPathChecklistValue } from '../types.ts';

export const checkFieldPathChecklist = (
  fieldPathChecklistValues: FieldPathChecklistValue[] | undefined,
  fieldTypeName: string | undefined,
): CheckFieldPathChecklistResult => {
  if (!fieldPathChecklistValues || fieldPathChecklistValues.length === 0) {
    return { hasData: false, typeUnused: !!fieldTypeName };
  }

  const [fieldPathChecklistValue] = fieldPathChecklistValues;

  if (fieldPathChecklistValue) {
    const { hasData, typeName } = fieldPathChecklistValue;
    const typeUnused = typeName ? typeName !== fieldTypeName : undefined;
    return { hasData, typeUnused };
  }

  return {
    hasData: fieldPathChecklistValues.some(({ hasData, typeName }) => typeName === fieldTypeName && hasData),
    typeUnused: !fieldPathChecklistValues.every(({ typeName }) => typeName === fieldTypeName),
  };
};
