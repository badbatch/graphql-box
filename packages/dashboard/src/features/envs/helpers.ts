import { type Step } from '../../types.ts';

export const sortSteps = (payload: Record<string, Step>) => (idA: string, idB: string) => {
  const stepA = payload[idA]!;
  const stepB = payload[idB]!;

  if (stepA.logOrder < stepB.logOrder) {
    return -1;
  }

  if (stepA.logOrder > stepB.logOrder) {
    return 1;
  }

  return 0;
};
