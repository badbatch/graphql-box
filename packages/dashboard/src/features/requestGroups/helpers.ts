import { type Env } from '../../types.ts';

export const sortEnvs = (payload: Record<string, Env>) => (idA: string, idB: string) => {
  const envA = payload[idA]!;
  const envB = payload[idB]!;

  if (envA.logGroup < envB.logGroup) {
    return -1;
  }

  if (envA.logGroup > envB.logGroup) {
    return 1;
  }

  return 0;
};
