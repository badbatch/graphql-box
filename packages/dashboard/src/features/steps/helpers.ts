import { type LogEntry } from '../../types.ts';

export const sortEntries = (payload: Record<string, LogEntry>) => (idA: string, idB: string) => {
  const logEntryA = payload[idA]!;
  const logEntryB = payload[idB]!;
  const timestampA = new Date(logEntryA['@timestamp']).valueOf();
  const timestampB = new Date(logEntryB['@timestamp']).valueOf();

  if (timestampA < timestampB) {
    return -1;
  }

  if (timestampA > timestampB) {
    return 1;
  }

  return 0;
};
