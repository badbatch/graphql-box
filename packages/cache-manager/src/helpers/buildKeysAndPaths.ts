import { FieldTypeInfo, PlainObjectMap } from "@graphql-box/core";
import { getAlias, getArguments, getName, hashRequest } from "@graphql-box/helpers";
import { FieldNode } from "graphql";
import { isNumber } from "lodash";
import { CacheManagerContext, KeysAndPaths, KeysAndPathsOptions } from "../defs";

export const buildKey = (path: string, key: string | number) => {
  const paths: (string | number)[] = [];

  if (path.length) {
    paths.push(path);
  }

  paths.push(key);
  return paths.join(".");
};

export const buildRequestFieldCacheKey = (
  name: string,
  requestFieldCacheKey: string,
  args: PlainObjectMap | undefined,
  directives?: FieldTypeInfo["directives"],
  index?: number,
) => {
  let key = `${isNumber(index) ? index : name}`;

  if (args) {
    key = `${key}(${JSON.stringify(args)})`;
  }

  if (directives?.inherited?.length) {
    key = `${directives.inherited.join(" ")} ${key}`;
  }

  if (directives?.own?.length) {
    key = `${key} ${directives.own.join(" ")}`;
  }

  return buildKey(requestFieldCacheKey, key);
};

export const buildFieldKeysAndPaths = (
  field: FieldNode,
  options: KeysAndPathsOptions,
  context: CacheManagerContext,
): KeysAndPaths => {
  const name = getName(field) as FieldNode["name"]["value"];
  const { index, requestFieldCacheKey = "", requestFieldPath = "", responseDataPath = "" } = options;
  const fieldAliasOrName = getAlias(field) || name;
  const updatedRequestFieldPath = isNumber(index) ? requestFieldPath : buildKey(requestFieldPath, fieldAliasOrName);
  const fieldTypeInfo = context.fieldTypeMap.get(updatedRequestFieldPath);

  const updatedRequestFieldCacheKey = buildRequestFieldCacheKey(
    name,
    requestFieldCacheKey,
    getArguments(field),
    fieldTypeInfo?.directives,
    index,
  );

  const propNameOrIndex = isNumber(index) ? index : fieldAliasOrName;
  const updatedResponseDataPath = buildKey(responseDataPath, propNameOrIndex);

  return {
    hashedRequestFieldCacheKey: hashRequest(updatedRequestFieldCacheKey),
    propNameOrIndex,
    requestFieldCacheKey: updatedRequestFieldCacheKey,
    requestFieldPath: updatedRequestFieldPath,
    responseDataPath: updatedResponseDataPath,
  };
};
