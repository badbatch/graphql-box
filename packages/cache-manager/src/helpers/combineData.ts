import { isObjectLike } from '@graphql-box/helpers';
import { mergeDataSets } from './mergeObjects.ts';

export const combineDataSets = (dataSetA: unknown, dataSetB: unknown, typeIDKey: string) => {
  if (!dataSetA && dataSetB) {
    return dataSetB;
  }

  if (isObjectLike(dataSetA) && isObjectLike(dataSetB)) {
    return mergeDataSets(dataSetA, dataSetB, typeIDKey);
  }

  return dataSetA;
};
