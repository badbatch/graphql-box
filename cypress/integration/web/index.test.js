import Cachemap from '@cachemap/core';
import indexedDB from '@cachemap/indexed-db';
import cacheManager from '@handl/cache-manager';
import Client from '@handl/client';
import { DEFAULT_TYPE_ID_KEY } from '@handl/core';
import debugManager from '@handl/debug-manager';
import fetchManager from '@handl/fetch-manager';
import { dehydrateCacheMetadata, hashRequest } from '@handl/helpers';
import requestParser from '@handl/request-parser';
import { githubIntrospection, parsedRequests, responses } from '@handl/test-utils';
import fetchMock from 'fetch-mock';

const defaultOptions = { awaitDataCaching: true, returnCacheMetadata: true };
const { performance } = window;
const url = 'https://api.github.com/graphql';

function log(...args) {
  console.log(...args); // eslint-disable-line no-console
}

async function initClient({ typeCacheDirectives }) {
  return Client.init({
    cacheManager: cacheManager({
      cache: await Cachemap.init({
        name: 'cachemap',
        store: indexedDB(),
      }),
      cascadeCacheControl: true,
      typeCacheDirectives,
    }),
    debugManager: debugManager({
      logger: { log },
      name: 'CLIENT',
      performance,
    }),
    requestManager: fetchManager({ url }),
    requestParser: requestParser({ introspection: githubIntrospection }),
    typeIDKey: DEFAULT_TYPE_ID_KEY,
  });
}

function buildRequestURL(hash) {
  if (!hash) return '*';
  return `${url}?requestId=${hash}`;
}

function mockRequest({ data, hash }) {
  const body = { data };
  const headers = { 'cache-control': 'public, max-age=5' };
  fetchMock.post(buildRequestURL(hash), { body, headers });
}

describe('@handl/client', () => {
  const realDateNow = Date.now.bind(window.Date);

  before(() => {
    window.Date.now = cy.stub().returns(Date.parse('June 6, 1979'));
  });

  after(() => {
    window.Date.now = realDateNow;
  });

  describe('request', () => {
    let cache, client, response;

    describe('no match', () => {
      before(async () => {
        mockRequest({ data: responses.singleTypeQuery.data });

        const typeCacheDirectives = {
          Organization: 'public, max-age=1',
        };

        try {
          client = await initClient({ typeCacheDirectives });
        } catch (errors) {
          log(errors);
        }

        const request = parsedRequests.singleTypeQuery;

        try {
          const { _cacheMetadata, data } = await client.request(request, { ...defaultOptions });
          response = { _cacheMetadata: dehydrateCacheMetadata(_cacheMetadata), data };
        } catch (errors) {
          log(errors);
        }

        cache = await client.cache.export();
      });

      after(async () => {
        fetchMock.restore();
        await client.cache.clear();
      });

      it('correct response data', () => {
        cy.wrap(response).toMatchSnapshot();
      });

      it('correct cache data', () => {
        cy.wrap(cache).toMatchSnapshot();
      });
    });

    describe('query tracker match', () => {
      before(async () => {
        mockRequest({ data: responses.singleTypeQuery.data });

        const typeCacheDirectives = {
          Organization: 'public, max-age=1',
        };

        try {
          client = await initClient({ typeCacheDirectives });
        } catch (errors) {
          log(errors);
        }

        const request = parsedRequests.singleTypeQuery;

        try {
          const result = await Promise.all([
            client.request(request, { ...defaultOptions }),
            client.request(request, { ...defaultOptions }),
          ]);

          const { _cacheMetadata, data } = result[1];
          response = { _cacheMetadata: dehydrateCacheMetadata(_cacheMetadata), data };
        } catch (errors) {
          log(errors);
        }

        cache = await client.cache.export();
      });

      after(async () => {
        fetchMock.restore();
        await client.cache.clear();
      });

      it('one request', () => {
        expect(fetchMock.calls()).to.have.lengthOf(1);
      });

      it('correct response data', () => {
        cy.wrap(response).toMatchSnapshot();
      });

      it('correct cache data', () => {
        cy.wrap(cache).toMatchSnapshot();
      });
    });

    describe('query response match', () => {
      before(async () => {
        mockRequest({ data: responses.singleTypeQuery.data });

        const typeCacheDirectives = {
          Organization: 'public, max-age=1',
        };

        try {
          client = await initClient({ typeCacheDirectives });
        } catch (errors) {
          log(errors);
        }

        const request = parsedRequests.singleTypeQuery;

        try {
          await client.request(request, { ...defaultOptions });
        } catch (errors) {
          log(errors);
        }

        fetchMock.resetHistory();

        try {
          const { _cacheMetadata, data } = await client.request(request, { ...defaultOptions });
          response = { _cacheMetadata: dehydrateCacheMetadata(_cacheMetadata), data };
        } catch (errors) {
          log(errors);
        }

        cache = await client.cache.export();
      });

      after(async () => {
        fetchMock.restore();
        await client.cache.clear();
      });

      it('no request', () => {
        expect(fetchMock.calls()).to.have.lengthOf(0);
      });

      it('correct response data', () => {
        cy.wrap(response).toMatchSnapshot();
      });

      it('correct cache data', () => {
        cy.wrap(cache).toMatchSnapshot();
      });
    });

    describe('request field path / data entity match', () => {
      before(async () => {
        mockRequest({ data: responses.singleTypeQuery.data });

        const typeCacheDirectives = {
          Organization: 'public, max-age=1',
        };

        try {
          client = await initClient({ typeCacheDirectives });
        } catch (errors) {
          log(errors);
        }

        const { full, initial } = parsedRequests.singleTypeQuerySet;

        try {
          await client.request(full, { ...defaultOptions });
        } catch (errors) {
          log(errors);
        }

        fetchMock.resetHistory();

        try {
          const { _cacheMetadata, data } = await client.request(initial, { ...defaultOptions });
          response = { _cacheMetadata: dehydrateCacheMetadata(_cacheMetadata), data };
        } catch (errors) {
          log(errors);
        }

        cache = await client.cache.export();
      });

      after(async () => {
        fetchMock.restore();
        await client.cache.clear();
      });

      it('no request', () => {
        expect(fetchMock.calls()).to.have.lengthOf(0);
      });

      it('correct response data', () => {
        cy.wrap(response).toMatchSnapshot();
      });

      it('correct cache data', () => {
        cy.wrap(cache).toMatchSnapshot();
      });
    });

    describe('request field path / data entity partial match', () => {
      before(async () => {
        const { full, initial, updated } = parsedRequests.nestedTypeQuerySet;

        mockRequest({
          data: responses.nestedTypeQuerySet.initial.data,
          hash: hashRequest(initial),
        });

        mockRequest({
          data: responses.nestedTypeQuerySet.updated.data,
          hash: hashRequest(updated),
        });

        const typeCacheDirectives = {
          Organization: 'public, max-age=3',
          Repository: 'public, max-age=3',
          RepositoryConnection: 'public, max-age=1',
          RepositoryOwner: 'public, max-age=3',
        };

        try {
          client = await initClient({ typeCacheDirectives });
        } catch (errors) {
          log(errors);
        }

        try {
          await client.request(initial, { ...defaultOptions });
        } catch (errors) {
          log(errors);
        }

        fetchMock.resetHistory();

        try {
          const { _cacheMetadata, data } = await client.request(full, { ...defaultOptions });
          response = { _cacheMetadata: dehydrateCacheMetadata(_cacheMetadata), data };
        } catch (errors) {
          log(errors);
        }

        cache = await client.cache.export();
      });

      after(async () => {
        fetchMock.restore();
        await client.cache.clear();
      });

      it('one request', () => {
        expect(fetchMock.calls()).to.have.lengthOf(1);
      });

      it('correct response data', () => {
        cy.wrap(response).toMatchSnapshot();
      });

      it('correct cache data', () => {
        cy.wrap(cache).toMatchSnapshot();
      });
    });
  });
});
