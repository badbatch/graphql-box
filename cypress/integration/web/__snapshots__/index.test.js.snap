exports[`@handl/client > request > no match > correct response data #0`] =
{
  "_cacheMetadata": {
    "query": {
      "cacheControl": {
        "maxAge": 1,
        "public": true
      },
      "ttl": 297471601000
    },
    "query.organization": {
      "cacheControl": {
        "maxAge": 1,
        "public": true
      },
      "ttl": 297471601000
    }
  },
  "data": {
    "organization": {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook"
    }
  }
};

exports[`@handl/client > request > no match > correct cache data #0`] =
{
  "entries": [
    [
      "dataEntities::Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      {
        "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
        "email": "",
        "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "login": "facebook",
        "name": "Facebook",
        "url": "https://github.com/facebook"
      }
    ],
    [
      "queryResponses::ea1cbc58eca798225b53f59d95ffb23a",
      {
        "cacheMetadata": {
          "query": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          },
          "query.organization": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          }
        },
        "data": {
          "organization": {
            "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
            "email": "",
            "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
            "login": "facebook",
            "name": "Facebook",
            "url": "https://github.com/facebook"
          }
        }
      }
    ],
    [
      "requestFieldPaths::5256db516a2438a7e5488a00413f4809",
      {
        "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
        "email": "",
        "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "login": "facebook",
        "name": "Facebook",
        "url": "https://github.com/facebook"
      }
    ]
  ],
  "metadata": [
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "requestFieldPaths::5256db516a2438a7e5488a00413f4809",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 410,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "dataEntities::Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 410,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "queryResponses::ea1cbc58eca798225b53f59d95ffb23a",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 678,
      "tags": [],
      "updatedCount": 0
    }
  ]
};

exports[`@handl/client > request > query response match > correct response data #0`] =
{
  "_cacheMetadata": {
    "query": {
      "cacheControl": {
        "maxAge": 1,
        "public": true
      },
      "ttl": 297471601000
    },
    "query.organization": {
      "cacheControl": {
        "maxAge": 1,
        "public": true
      },
      "ttl": 297471601000
    }
  },
  "data": {
    "organization": {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "email": "",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook",
      "name": "Facebook",
      "url": "https://github.com/facebook"
    }
  }
};

exports[`@handl/client > request > query response match > correct cache data #0`] =
{
  "entries": [
    [
      "dataEntities::Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      {
        "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
        "email": "",
        "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "login": "facebook",
        "name": "Facebook",
        "url": "https://github.com/facebook"
      }
    ],
    [
      "queryResponses::ea1cbc58eca798225b53f59d95ffb23a",
      {
        "cacheMetadata": {
          "query": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          },
          "query.organization": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          }
        },
        "data": {
          "organization": {
            "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
            "email": "",
            "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
            "login": "facebook",
            "name": "Facebook",
            "url": "https://github.com/facebook"
          }
        }
      }
    ],
    [
      "requestFieldPaths::5256db516a2438a7e5488a00413f4809",
      {
        "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
        "email": "",
        "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "login": "facebook",
        "name": "Facebook",
        "url": "https://github.com/facebook"
      }
    ]
  ],
  "metadata": [
    {
      "accessedCount": 1,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "queryResponses::ea1cbc58eca798225b53f59d95ffb23a",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 678,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "requestFieldPaths::5256db516a2438a7e5488a00413f4809",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 410,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "dataEntities::Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 410,
      "tags": [],
      "updatedCount": 0
    }
  ]
};

exports[`@handl/client > request > request field path / data entity match > correct response data #0`] =
{
  "_cacheMetadata": {
    "query": {
      "cacheControl": {
        "maxAge": 1,
        "public": true
      },
      "ttl": 297471601000
    },
    "query.organization": {
      "cacheControl": {
        "maxAge": 1,
        "public": true
      },
      "ttl": 297471601000
    }
  },
  "data": {
    "organization": {
      "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
      "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "login": "facebook"
    }
  }
};

exports[`@handl/client > request > request field path / data entity match > correct cache data #0`] =
{
  "entries": [
    [
      "dataEntities::Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      {
        "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
        "email": "",
        "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "login": "facebook",
        "name": "Facebook",
        "url": "https://github.com/facebook"
      }
    ],
    [
      "queryResponses::ea1cbc58eca798225b53f59d95ffb23a",
      {
        "cacheMetadata": {
          "query": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          },
          "query.organization": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          }
        },
        "data": {
          "organization": {
            "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
            "email": "",
            "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
            "login": "facebook",
            "name": "Facebook",
            "url": "https://github.com/facebook"
          }
        }
      }
    ],
    [
      "queryResponses::f666cc27cdcfcecacb340b64fcf27506",
      {
        "cacheMetadata": {
          "query": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          },
          "query.organization": {
            "cacheControl": {
              "maxAge": 1,
              "public": true
            },
            "ttl": 297471601000
          }
        },
        "data": {
          "organization": {
            "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
            "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
            "login": "facebook"
          }
        }
      }
    ],
    [
      "requestFieldPaths::5256db516a2438a7e5488a00413f4809",
      {
        "description": "We are working to build community through open source technology. NB: members must have two-factor auth.",
        "email": "",
        "id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
        "login": "facebook",
        "name": "Facebook",
        "url": "https://github.com/facebook"
      }
    ]
  ],
  "metadata": [
    {
      "accessedCount": 1,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "requestFieldPaths::5256db516a2438a7e5488a00413f4809",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 410,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 1,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "dataEntities::Organization::MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 410,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "queryResponses::f666cc27cdcfcecacb340b64fcf27506",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 584,
      "tags": [],
      "updatedCount": 0
    },
    {
      "accessedCount": 0,
      "added": 297471600000,
      "cacheability": {
        "metadata": {
          "cacheControl": {
            "maxAge": 1,
            "public": true
          },
          "ttl": 297471601000
        }
      },
      "key": "queryResponses::ea1cbc58eca798225b53f59d95ffb23a",
      "lastAccessed": 297471600000,
      "lastUpdated": 297471600000,
      "size": 678,
      "tags": [],
      "updatedCount": 0
    }
  ]
};
